internal/rpc/digest_md5.go

```diff
@@ -0,0 +1,426 @@
+func (d *digestMD5) a1() string {
+	return a1(string(d.authID), d.token.realm, d.passwd, d.token.nonce, d.cnonce)
+}
+
+// convenience function for calling the a2 function for evaluating challenges
+func (d *digestMD5) a2(first bool) string {
+	return a2(d.service+"/"+d.hostname, d.token.qop, first)
+}
+
+// challengeStep1 implements Using Digest Authentication as a SASAL Mechanism
+// as per RFC2831.
+func (d *digestMD5) challengeStep1(challenge []byte) (string, error) {
+	ncval := fmt.Sprintf("%08x", 1)
+	var err error
+	d.token, err = parseChallenge(challenge)
+	if err != nil {
+		return "", err
+	}
+
+	d.cnonce, err = genCnonce()
+	if err != nil {
+		return "", err
+	}
+
+	d.cipher = chooseCipher(d.token.cipher)
+	rspdigest := compute(d.a1(), d.a2(true), d.token.nonce, d.cnonce, ncval, d.token.qop)
+
+	ret := fmt.Sprintf(`username="%s", realm="%s", nonce="%s", cnonce="%s", nc=%08x, qop=%s, digest-uri="%s/%s", response=%s, charset=utf-8`,
+		d.authID, d.token.realm, d.token.nonce, d.cnonce, 1, d.token.qop, d.service, d.hostname, rspdigest)
+
+	if d.cipher != cipherUnknown {
+		ret += ", cipher=" + d.cipher.String()
+	}
+	return ret, nil
+}
+
+// challengeStep2 implements Using Digest Authentication as a SASAL Mechanism
+// as per RFC2831.
+func (d *digestMD5) challengeStep2(challenge []byte) (string, error) {
+	ncval := fmt.Sprintf("%08x", 1)
+	rspdigest := compute(d.a1(), d.a2(false), d.token.nonce, d.cnonce, ncval, d.token.qop)
+
+	rspauth := strings.Split(string(challenge), "=")
+	if rspauth[0] != "rspauth" {
+		return "", fmt.Errorf("Could not find rspauth in '%s'", string(challenge))
+	}
+	if rspauth[1] != rspdigest {
+		return "", fmt.Errorf("rspauth did not match digest")
+	}
+	return "", nil
+}
+
+// make this easy to mock out for testing by having it defined here where
+// unit tests can override it
+var genCnonce = func() (string, error) {
+	ret := make([]byte, 12)
+	if _, err := rand.Read(ret); err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(ret), nil
+}
+
+// newDigestMD5Conn performs the necessary handshake based on the configuration and then returns
+// a wrapped connection or error, in the case of our auth set to only authentication, then it just
+// returns the connection unmodified after authenticating
+func newDigestMD5Conn(conn net.Conn, token *hadoop_common.TokenProto, key *EncryptionKey, useSecure bool) (net.Conn, error) {
+	auth := &hadoop_common.RpcSaslProto_SaslAuth{}
+	auth.Method = &authMethod
+	auth.Mechanism = &authMechanism
+	auth.ServerId = &authServer
+	auth.Protocol = &authProtocol
+
+	ourToken := &hadoop_common.TokenProto{}
+	ourToken.Kind = token.Kind
+	ourToken.Password = token.Password[:]
+	ourToken.Service = token.Service
+	ourToken.Identifier = token.GetIdentifier()
+
+	// if the server didn't send us a Nonce, then the data isn't encrypted
+	// but we will still attempt to authenticate
+	if useSecure && len(key.Nonce) == 0 {
```

Comment:
In which case does the server not send a nonce?

Reply:
if the client has the data transfer protection set to privacy, but the server is configured for auth-integrity or authentication but not auth-privacy, then there is no key to send and as such the nonce will be empty.  In that case we fall back to what the server's qop is.